roachpb: InternalServer API for tenant settings

RaduBerinde opened the pull request:
This commit introduces the API that the tenants will use to obtain and
list for updates to tenant setting overrides. The API was designed to
allow for maximum flexibility on the server side so that it can be
implemented as a range feed without any extra state (if necessary).

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74268)
<!-- Reviewable:end -->

RaduBerinde commented:
Friendly ping.

ajwerner reviewed (commented):
High level LGTM

> Reviewed 1 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74268#-:-Mt4AMN93w6PwdmBjj62:b1dz7ih)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @RaduBerinde)

___
*[pkg/roachpb/api.proto, line 2631 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74268#-Mt49JZh1BjLIigyZjXk:-Mt49JZi3Hyfn6OE0gag:b3gfblb) ([raw file](https://github.com/cockroachdb/cockroach/blob/26a895399b2772243ff9b034eb9e61b636b5c392/pkg/roachpb/api.proto#L2631)):*
> ```protobuf
> // server side (e.g. to make it maintain very little state per tenant).
> message TenantSettingsEvent {
>   // Tier indicates the tier and is either 1 or 2.
> ```

Any interest in a `gogoproto.casttype`? 



_Code quote:_
```protobuf
  // Tier indicates the tier and is either 1 or 2.
  uint32 tier = 1;
```
___
*[pkg/roachpb/api.proto, line 2646 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74268#-Mt49RWPAKjuUWpAdV0Y:-Mt49RWPAKjuUWpAdV0Z:b-2hj2cp) ([raw file](https://github.com/cockroachdb/cockroach/blob/26a895399b2772243ff9b034eb9e61b636b5c392/pkg/roachpb/api.proto#L2646)):*
> ```protobuf
>   // Overrides contains:
>   //  - all current setting overrides for the given tier if Incremental is
>   //    false; or 
> ```

nit: trailing spaces


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
roachpb: InternalServer API for tenant settings

This commit introduces the API that the tenants will use to obtain and
list for updates to tenant setting overrides. The API was designed to
allow for maximum flexibility on the server side so that it can be
implemented as a range feed without any extra state (if necessary).

Release note: None

RaduBerinde reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74268#-:-Mt4K6NVBYLdT5mikWCy:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @dt)

___
*[pkg/roachpb/api.proto, line 2631 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74268#-Mt49JZh1BjLIigyZjXk:-Mt4JR6a6JCAozNXJD2B:b-6hro0d) ([raw file](https://github.com/cockroachdb/cockroach/blob/26a895399b2772243ff9b034eb9e61b636b5c392/pkg/roachpb/api.proto#L2631)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Any interest in a `gogoproto.casttype`? 
</blockquote></details>

Good idea, done. While writing this I also thought about whether `precedence` would be more descriptive than `tier`.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
test

dt reviewed (approved):
I like "precedence" too

RaduBerinde reviewed (commented):
> I like "precedence" too

Done! TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74268#-:-Mt4_MyE67KYgNVhWeNv:b-os5o79)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/roachpb/api.proto, line 2651 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/74268#-Mt4_U2SB9BWBXdf73Er:-Mt4_U2SB9BWBXdf73Es:b-aopzpi) ([raw file](https://github.com/cockroachdb/cockroach/blob/63fd95ff352dc5f5d86cd02fd6526c622d16d983/pkg/roachpb/api.proto#L2651)):*
> ```protobuf
>   //    fields).
>   repeated TenantSetting overrides = 3;
> ```

Moved this error field, I must have put it inside `TenantSetting` by accident.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74268#-:-Mt4nQGX0otJSHf6ZZhJ:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @RaduBerinde)

___
*[pkg/roachpb/api.proto, line 2631 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74268#-Mt49JZh1BjLIigyZjXk:-Mt4nMKL6DN9nHSdYSvA:b-hk0r2p) ([raw file](https://github.com/cockroachdb/cockroach/blob/26a895399b2772243ff9b034eb9e61b636b5c392/pkg/roachpb/api.proto#L2631)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Good idea, done. While writing this I also thought about whether `precedence` would be more descriptive than `tier`.
</blockquote></details>

Also +1 on `precedence` 


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+
